File: 49.group-anagrams.py
#
# @lc app=leetcode id=49 lang=python3
#
# [49] Group Anagrams
#

# @lc code=start
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = {}

        # Go through each string and sort it, then use that to store anagrams together in dictionary
        # where key is sorted anagram, and value is list of grouped anagrams
        for string in strs:
            anagram = "".join(sorted(string))
            if anagram not in anagrams: anagrams[anagram] = []
            anagrams[anagram].append(string)

        # Add subarrays to res
        res = []
        for val in anagrams.values():
            res.append(val)
        
        return res
        
# @lc code=end

File: 49.group-anagrams.cpp
/*
 * @lc app=leetcode id=49 lang=cpp
 *
 * [49] Group Anagrams
 */

// @lc code=start
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> result;
        unordered_map<string, vector<string>> map;

        for (string word : strs) {
            string key = word;
            sort(key.begin(), key.end());
            map[key].push_back(word);
        }

        for (const pair<const string, vector<string>>& pair : map) {
            result.push_back(pair.second);
        }

        return result;
    }
};
// @lc code=end

